massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1074 @@
+use std::cmp::Ordering;
+use std::collections::{BTreeMap, BTreeSet, HashMap};
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::sync::Arc;
+
+use machine::{machine, transitions};
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use parking_lot::RwLock;
+use thiserror::Error;
+
+use massa_models::{amount::Amount, config::VERSIONING_THRESHOLD_TRANSITION_ACCEPTED};
+use massa_time::MassaTime;
+
+// TODO: add more items here
+/// Versioning component enum
+#[allow(missing_docs)]
+#[derive(Clone, Debug, PartialEq, Eq, Hash, TryFromPrimitive, IntoPrimitive)]
+#[repr(u32)]
+pub enum MipComponent {
+    Address,
+    Block,
+    VM,
+}
+
+/// MIP info (name,  version,
+#[derive(Clone, Debug)]
+pub struct MipInfo {
+    /// MIP name or descriptive name
+    pub name: String,
+    /// Network (or global) version (to be included in block header)
+    pub version: u32,
+    /// Component concerned by this versioning (e.g. a new Block version)
+    pub component: MipComponent,
```

Comment:
For now, yes we can only have one component update by version. An improvement can be found here: https://github.com/massalabs/massa/issues/3631

Reply:
Thinking about this, I think it may be better to change:
```rust
pub component: MipComponent,
pub component_version: u64,
```
to
```rust
pub components: HashMap<MipComponent, u32>
```
Because we may need to update multiple components at once.

What BIP-9 proposes (and what would be in the follow-up) would be to decouple different component upgrades. E.g. you can vote to change both the Bytecode version and the Address version even though they are unrelated.

Reply:
Do you think we need to do this in this PR or in the follow up? @Leo-Besancon

Reply:
So let's do it in the next PR :)